pkg/operator/controller/controller_dns_node_resolver_daemonset.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// services is a comma- or space-delimited list of services for which
+	// entries should be added to /etc/hosts.  NOTE: For now, ensure these
+	// are relative names; for each relative name, an alias with the
+	// CLUSTER_DOMAIN suffix will also be added.
+	services = "image-registry.openshift-image-registry.svc"
+)
+
+var (
+	// nodeResolverScript is a shell script that updates /etc/hosts.
+	nodeResolverScript = manifests.NodeResolverScript()
+)
+
+// ensureNodeResolverDaemonset ensures the node resolver daemonset exists if it
+// should or does not exist if it should not exist.  Returns a Boolean
+// indicating whether the daemonset exists, the daemonset if it does exist, and
+// an error value.
+func (r *reconciler) ensureNodeResolverDaemonSet(clusterIP, clusterDomain string) (bool, *appsv1.DaemonSet, error) {
+	haveDS, current, err := r.currentNodeResolverDaemonSet()
+	if err != nil {
+		return false, nil, err
+	}
+	wantDS, desired, err := desiredNodeResolverDaemonSet(clusterIP, clusterDomain, r.OpenshiftCLIImage)
+	if err != nil {
+		return haveDS, current, fmt.Errorf("failed to build node resolver daemonset: %v", err)
+	}
+	switch {
+	case !wantDS && !haveDS:
+		return false, nil, nil
+	case !wantDS && haveDS:
+		if err := r.deleteNodeResolverDaemonSet(current); err != nil {
+			return true, current, err
+		}
+		return false, nil, nil
+	case wantDS && !haveDS:
+		if err := r.createNodeResolverDaemonSet(desired); err != nil {
+			return false, nil, err
+		}
+		return r.currentNodeResolverDaemonSet()
+	case wantDS && haveDS:
+		if updated, err := r.updateNodeResolverDaemonSet(current, desired); err != nil {
+			return true, current, err
+		} else if updated {
+			return r.currentNodeResolverDaemonSet()
+		}
+	}
+	return true, current, nil
+}
+
+// desiredNodeResolverDaemonSet returns the desired node resolver daemonset.
+func desiredNodeResolverDaemonSet(clusterIP, clusterDomain, openshiftCLIImage string) (bool, *appsv1.DaemonSet, error) {
+	hostPathFile := corev1.HostPathFile
+	envs := []corev1.EnvVar{{
+		Name:  "SERVICES",
+		Value: services,
+	}}
+	if len(clusterIP) > 0 {
+		envs = append(envs, corev1.EnvVar{
+			Name:  "NAMESERVER",
+			Value: clusterIP,
+		})
+	}
+	if len(clusterDomain) > 0 {
+		envs = append(envs, corev1.EnvVar{
+			Name:  "CLUSTER_DOMAIN",
+			Value: clusterDomain,
+		})
+	}
+	trueVal := true
+	name := NodeResolverDaemonSetName()
+	daemonset := appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name.Name,
+			Namespace: name.Namespace,
+		},
+		Spec: appsv1.DaemonSetSpec{
+			Selector: NodeResolverDaemonSetPodSelector(),
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: NodeResolverDaemonSetPodSelector().MatchLabels,
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{{
```

Comment:
Would it be possible to use `hostNetwork: true` for the node resolver daemonset?

This is related to the Smart-NIC use case we are trying to address (WIP) on OpenShift baremetal, where for a particular node, it is not possible to create a regular veth pod with ovn-k8s network type. so every infra pod like dns resolver that needs to run on every node (including the Smart-NIC node) may not be able to create if not running with `hostNetwork:true`.

Reply:
I don't see why we couldn't use `hostNetwork: true` for the node resolver daemonset.  If I understand correctly, the use-case is that the daemonset's pods may need to run on nodes where container networking cannot be configured for the pod; in this scenario, can the pod still use the service network to preform DNS queries?

Reply:
> I don't see why we couldn't use `hostNetwork: true` for the node resolver daemonset. If I understand correctly, the use-case is that the daemonset's pods may need to run on nodes where container networking cannot be configured for the pod; 

Container networking can only be used together with a SR-IOV (hardware device) interface in Smart-NIC case.
Since SR-IOV devices are limited resource on a k8s node, resources.limits & resources.requests are required in pod spec in order for dns or other infra pods to request such network (other pod spec changes like pod annotation might also be needed). So an easy fix in my opinion is to run dns pod in hostNetwork - not using container networking, Or be able to update the dns pod spec with a user-defined resource which I think is not trivial.

> in this scenario, can the pod still use the service network to perform DNS queries?

Yes, service network will still be available.

Reply:
Yeah, that makes sense.  I don't see any problem with using `hostNetwork: true`, but I'll raise the question with my team for feedback.

Reply:
Latest push specifies `hostNetwork: false` on the node-resolver daemonset.

Reply:
> Latest push specifies `hostNetwork: false` on the node-resolver daemonset.

Thanks @Miciah ! 
I think you mean `hostNetwork: true` as written in the code :-)

Reply:
> I think you mean `hostNetwork: true` as written in the code :-)

Indeed!  The code was right, but the GitHub comment was wrong.  Sorry for the error!